core/src/Kokkos_MathematicalFunctions.hpp

```diff
@@ -0,0 +1,232 @@
+    defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ENABLE_OPENMPTARGET)
+
+#define KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)                         \
+  template <typename Arithmetic1, typename Arithmetic2,                      \
+            typename = std::enable_if_t<                                     \
+                std::is_arithmetic<Arithmetic1>::value &&                    \
+                std::is_arithmetic<Arithmetic2>::value &&                    \
+                !std::is_same<Arithmetic1, long double>::value &&            \
+                !std::is_same<Arithmetic2, long double>::value>>             \
+  KOKKOS_INLINE_FUNCTION double FUNC(Arithmetic1 x, Arithmetic2 y) {         \
+    return Kokkos::Experimental::FUNC(                                       \
+        static_cast<std::conditional_t<std::is_integral<Arithmetic1>::value, \
+                                       double, Arithmetic1>>(x),             \
+        static_cast<std::conditional_t<std::is_integral<Arithmetic2>::value, \
+                                       double, Arithmetic2>>(y));            \
+  }
+
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC)                     \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, float, float)   \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, double, double) \
+  KOKKOS_IMPL_UNARY_FUNCTION_INTEGRAL(FUNC, double)
+
+#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC)                  \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, bool, float)  \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, bool, double) \
+  KOKKOS_IMPL_UNARY_FUNCTION_INTEGRAL(FUNC, bool)
+
+#define KOKKOS_IMPL_MATH_BINARY_FUNCTION(FUNC)             \
+  KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, float)  \
+  KOKKOS_IMPL_BINARY_FUNCTION_FLOATING_POINT(FUNC, double) \
+  KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)
+
+#define KOKKOS_IMPL_MATH_NAN()                                        \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nanf, float, char const*) \
+  KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(nan, double, char const*)
+
+#else  // long double overloads are available
+
+#define KOKKOS_IMPL_BINARY_FUNCTION_ARITHMETIC(FUNC)                         \
+  template <typename Arithmetic1, typename Arithmetic2,                      \
+            typename =                                                       \
+                std::enable_if_t<std::is_arithmetic<Arithmetic1>::value &&   \
+                                 std::is_arithmetic<Arithmetic2>::value>,    \
+            typename Promoted = std::conditional_t<                          \
+                std::is_same<Arithmetic1, long double>::value ||             \
+                    std::is_same<Arithmetic2, long double>::value,           \
+                long double, double>>                                        \
+  KOKKOS_INLINE_FUNCTION Promoted FUNC(Arithmetic1 x, Arithmetic2 y) {       \
```

Comment:
Same as above. Always returns either `long double` or `double`.

Reply:
This is a good question.  I am not sure whether implementations are allowed to specialize for other types of floating-point.

Reply:
In any case, we will tackle quadruple-precision in another PR.